### ISSUE TITLE
Feature request: support Babel 7

### ISSUE THREAD

**Issue by OKNoah** [codemix/flow-runtime#184]
This is a:

 - [x] Bug Report
 - [x] Feature Request

Which concerns:

 - [x] babel-plugin-flow-runtime

---

### What is the current behaviour?

**The complete configuration is here:** https://github.com/OKNoah/flow-runtime-test/pull/2/files

```json
{
  "presets": [
    "@babel/preset-env",
    "@babel/preset-stage-2"
  ],
  "plugins": [
    ["flow-runtime"],
    "transform-decorators-legacy"
  ]
}
```

```bash
/Users/noah/Projects/flow-runtime-test/node_modules/@babel/core/lib/transformation/normalize-file.js:136
    throw err;
    ^

SyntaxError: /Users/noah/Projects/flow-runtime-test/index.js: Unexpected token, expected ";" (3:5)

  1 | // @flow-runtime
  2 | 
> 3 | type User = {
    |      ^
  4 |   name: string
  5 | };
  6 | 
    at Parser.raise (/Users/noah/Projects/flow-runtime-test/node_modules/@babel/core/node_modules/babylon/lib/index.js:832:15)
    at Parser.unexpected (/Users/noah/Projects/flow-runtime-test/node_modules/@babel/core/node_modules/babylon/lib/index.js:2212:16)
    at Parser.semicolon (/Users/noah/Projects/flow-runtime-test/node_modules/@babel/core/node_modules/babylon/lib/index.js:2192:40)
    at Parser.parseExpressionStatement (/Users/noah/Projects/flow-runtime-test/node_modules/@babel/core/node_modules/babylon/lib/index.js:4729:10)
    at Parser.parseStatementContent (/Users/noah/Projects/flow-runtime-test/node_modules/@babel/core/node_modules/babylon/lib/index.js:4327:19)
    at Parser.parseStatement (/Users/noah/Projects/flow-runtime-test/node_modules/@babel/core/node_modules/babylon/lib/index.js:4206:17)
    at Parser.parseBlockOrModuleBlockBody (/Users/noah/Projects/flow-runtime-test/node_modules/@babel/core/node_modules/babylon/lib/index.js:4764:23)
    at Parser.parseBlockBody (/Users/noah/Projects/flow-runtime-test/node_modules/@babel/core/node_modules/babylon/lib/index.js:4750:10)
    at Parser.parseTopLevel (/Users/noah/Projects/flow-runtime-test/node_modules/@babel/core/node_modules/babylon/lib/index.js:4174:10)
    at Parser.parse (/Users/noah/Projects/flow-runtime-test/node_modules/@babel/core/node_modules/babylon/lib/index.js:5613:17)
[nodemon] app crashed - waiting for file changes before starting...
```

---

### What is the expected behaviour?

Should run without errors and do a console log.

---

### Which package versions are you using?

```json
"@babel/core": "^7.0.0-beta.36",
"@babel/preset-env": "^7.0.0-beta.36",
"@babel/preset-stage-2": "^7.0.0-beta.36",
"@babel/register": "^7.0.0-beta.36",
"babel-plugin-flow-runtime": "^0.15.0",
"babel-plugin-transform-decorators-legacy": "^1.3.4",
"flow-runtime": "^0.16.0"
```

**Comment by jedwards1211** [comment]
I don't know about anyone else, but I probably won't take a stab at Babel 7 support until it's out of beta.

**Comment by phpnode** [comment]
agree with @jedwards1211 - we obviously want to support 7 but I'm wary of being an early adopter

**Comment by JacopKane** [comment]
Currently, this is the only one preset that is stopping us migrating to babel 7.
Wish I've had some AST experience. I will try to check it out later if it's something I can handle

**Comment by billinghamj** [comment]
They have stated that it is effectively ready for production use and no significant breaking changes are to be expected.

**Comment by devcorpio** [comment]
Same problem here :(, we are trying to update from nextjs 5.1 to nextjs 6.0.3, but we are blocked too:

> Module build failed: SyntaxError: /components/Card/Card.js: Unexpected token, expected ";" (10:5)

   8 | import styles from './cardStyles';
   9 | 
> 10 | type Props = {
     |      ^
  11 |   title: string,

**Comment by jedwards1211** [comment]
@devcorpio nextjs takes that much control of the build pipeline that there's no way to override what it's using to transpile JS files?

**Comment by rej156** [comment]
Bump

**Comment by vjpr** [comment]
Started my own branch here: https://github.com/vjpr/flow-runtime/tree/vjpr/babel-7

**Comment by vjpr** [comment]
I traced down the exact code that is causing this snippet to fail.

```
export function foo() {
  {
    const bar: number = 1
  }
}
```

```
Error: TypeError: test.js: Duplicate declaration "bar" (This is an error on an internal node. Probably an internal error.)
```

https://github.com/codemix/flow-runtime/blob/8068da7c73225aad569fb46b6d981ddd6461ff4a/packages/babel-plugin-flow-runtime/src/transformVisitors.js#L276-L286

If a change `t.identifier(name)` to `t.identifier(name + 'blah')` the error is gone and the generated code looks fine.

jedwards1211 Could you shed some light on how this works?

**Comment by 0rvar** [comment]
Babel 7 has been released!

**Comment by bsmith-cycorp** [comment]
Any movement on this? I'm blocked by a couple of these "Duplicate declaration" errors.

**Comment by semireg** [comment]
> Seeing as it only occurs if we have a named export and an ES6 block statement, perhaps it is more of a babel@7 bug.
Here are the pertinent babel@7 issues currently open: https://github.com/babel/babel/issues/8899, https://github.com/babel/babel/issues/8525, https://github.com/babel/babel/issues/8559

I haven't dug in too deep, but after fixing https://github.com/codemix/flow-runtime/issues/216 I'm now seeing "Duplicate declarations" when I use the same bundling command.

The dupes I've seen so far seem to generate from destructuring parameters:
```
TypeError: /path/node_modules/react-native/Libraries/Interaction/TaskQueue.js: Duplicate declaration "onMoreTasks"
  50 |    * tasks to process.
  51 |    */
> 52 |   constructor({onMoreTasks}: {onMoreTasks: () => void}) {
     |                ^
  53 |     this._onMoreTasks = onMoreTasks;
  54 |     this._queueStack = [{tasks: [], popable: false}];
  55 |   }
    at File.buildCodeFrameError (/path/node_modules/@babel/core/lib/transformation/file/file.js:261:12)
    at Scope.checkBlockScopedCollisions (/path/node_modules/@babel/traverse/lib/scope/index.js:347:22)
    at Scope.registerBinding (/path/node_modules/@babel/traverse/lib/scope/index.js:504:16)
    at Scope.registerDeclaration (/path/node_modules/@babel/traverse/lib/scope/index.js:444:14)
    at Object.BlockScoped (/path/node_modules/@babel/traverse/lib/scope/index.js:189:28)
    at Object.newFn (/path/node_modules/@babel/traverse/lib/visitors.js:230:17)
    at NodePath._call (/path/node_modules/@babel/traverse/lib/path/context.js:53:20)
    at NodePath.call (/path/node_modules/@babel/traverse/lib/path/context.js:36:14)
    at NodePath.visit (/path/node_modules/@babel/traverse/lib/path/context.js:88:12)
```

Yet, I can't reproduce the failure in the tests. This passes fine:
```
export const input = `
function foo({bar}: {bar:string}) {
  console.log(bar);
}
`;

export const expected = `
import t from "flow-runtime";
function foo(_arg) {
  let {
    bar
  } = t.object(t.property("bar", t.string())).assert(_arg);
  console.log(bar);
}
`;
```

**Comment by jedwards1211** [comment]
Fixed in [v0.18.0](https://github.com/codemix/flow-runtime/releases/tag/v0.18.0)!
